add fixtures file in place of permissions.json

dscottboggs opened the pull request.

dscottboggs committed:
add fixtures file in place of permissions.json

j8r commented on spec/spec_helper.cr:
```diff
@@ -15,12 +12,26 @@ def new_test_context(verb = "GET", path = "/api/test")
+# the location
+macro permissions_file!
+  Path[__DIR__, "permissions.json"]
+end
```
The previous `PERMISSION_FILE` constant was better, a macro isn't needed here.

dscottboggs replied:
Assigning this to a constant broke things elsewhere due to `Path.[]` being a `Call` as opposed to `StringInterpolation` in macro-land. However, that issue seems to have cleared up.

j8r commented on spec/fixtures.cr:
```diff
@@ -0,0 +1,63 @@
+          name:         "Jim Oliver",
+        },
+      ],
+    }
```
Better, but I think we can further improve this: How this named tuple was created? This is too magic. I suppose the `DppmRestApi` was used at first, all we should need is only `DppmRestApi::Config`.
We may also have a problems if the JSON shema change.

dscottboggs committed:
Made fixtures less magic

j8r commented on spec/fixtures.cr:
```diff
@@ -0,0 +1,66 @@
+
+module DppmRestApi
+  struct Config
+    def initialize(@groups, @users)
```
Why? If this initializer is needed, other may need it too – so it can be added to `/src/config.cr`.

j8r commented on spec/fixtures.cr:
```diff
@@ -0,0 +1,66 @@
+require "../src/config"
+require "../src/ext/scrypt_password"
```
Not needed

dscottboggs committed:
replace macro with constant

dscottboggs committed:
polish and revise

the CLI spec no longer needs the with_state_restore;
moved Fixtures to the top-level;
removed reopening of Config in Fixtures

j8r commented on spec/spec_helper.cr:
```diff
@@ -15,12 +12,19 @@ def new_test_context(verb = "GET", path = "/api/test")
+Spec.before_each do
+  DppmRestApi.permissions_config = Fixtures.permissions_config
+  DppmRestApi.permissions_config.write_to PERMISSION_FILE
+  DppmRestApi.run Socket::IPAddress::LOOPBACK, DPPM::Prefix.default_dppm_config.port, __DIR__
```
You think it's better to create a new server for each spec? This will slow it down.
Reseting the permission file may be enough, with eventually reloading the server.

dscottboggs replied:
Generally, I prefer to make my tests as atomic and robust as possible, with little regard to the time it will take. However, in the case of these changes, you are right -- the slowdown is _significant_. I suspect the majority of that is from the time spent writing, deleting, and rewriting the file, but I'll do some checking and see if your idea will help significantly.

I was thinking earlier, perhaps instead of deleting the file every time, just rewrite it if the MD5 has does not match. This would be a read-only activity which could be cached by the Kernel. I will do some testing.

j8r reviewed (approved):
Nice! One more question.

j8r commented:
For some reason this branch can't be built in the CI – it was stuck with a High GC allocation count warning.

dscottboggs commented:
> For some reason this branch can't be built in the CI – it was stuck with a High GC allocation count warning.

probably due to the spec performance issue. The reason I had spun up a server in each spec was because the permissions_config.json has to be written before the server tries to start. Working on resolving those issues now.

dscottboggs commented:
Just checked it with the MD5 hash solution -- it increased the spec runtime from 26 to 32 seconds. :roll_eyes:

j8r commented:
~The CI is stuck on the build stage `shards build`.~ It's stuck at the end of `shards install`.
It's fine to let the `spec_helper` as is, the most inportant, as you said, is to have individual, atomic tests.

j8r commented:
After digging, good chances to be related to the [recent update of shards in Alpine to 0.9.0]( https://github.com/alpinelinux/aports/blob/master/community/shards/APKBUILD), done 2 days ago.

j8r commented:
Got it: https://github.com/crystal-lang/shards/issues/282#issuecomment-503152799

dscottboggs commented:
Interesting. Is there any reason not to merge in that case?
